plantio/serializers/aplicacao.py

```diff
@@ -2,32 +2,58 @@
 
 from plantio.models import AplicacaoAgrotoxico
 
+from core.consts.agrotoxicos import AplicacaoEstados
+
 from rest_framework import serializers
 
 
-# TODO: As linhas comentadas fazem referência a implementação da issue 152
 class AplicacaoAgrotoxicoSerializer(serializers.ModelSerializer):
-    # fotoAgrotoxico = serializers
-
     class Meta:
         model = AplicacaoAgrotoxico
         fields = (
-            # 'idAgrotoxico', 'dataAplicacao', 'fotoAgrotoxico', 'dosagemAplicacao'
-            'plantio', 'agrotoxico', 'dataAplicacao', 'dosagemAplicacao'
+            'plantio', 'agrotoxico', 'dataAplicacao', 'dosagemAplicacao', 'fotoAgrotoxico', 'estadoAnalise'
         )
+        read_only_fields = ('estadoAnalise',)
+        extra_kwargs = {
+            'agrotoxico': {'required': False},
+            'fotoAgrotoxico': {'required': False}
+        }
+
+    def create(self, validated_data):
+        estadoAnalise = AplicacaoEstados.SUCESSO
+
+        if validated_data.get("fotoAgrotoxico"):
+            estadoAnalise = AplicacaoEstados.ANALISE
+
+        return AplicacaoAgrotoxico.objects.create(estadoAnalise=estadoAnalise, **validated_data)
 
     def validate_dataAplicacao(self, dataAplicacao):
         if dataAplicacao > date.today():
             raise serializers.ValidationError('Data de aplicação no futuro.')
         return dataAplicacao
 
     def validate(self, data):
-        if data.get('agrotoxico'):
-            query = AplicacaoAgrotoxico.objects.filter(
-                plantio=data['plantio'], agrotoxico=data['agrotoxico'], dataAplicacao=data['dataAplicacao'])
 
-            if query.exists():
-                raise serializers.ValidationError(
-                    f'Esse agrotóxico já foi aplicado nessa plantação na data {data["dataAplicacao"]}.')
+        self._validate_agrotoxico_xor_fotoAgrotoxico(data.get('agrotoxico'), data.get('fotoAgrotoxico'))
+        self._validate_agrotoxico(data.get('agrotoxico'), data['plantio'], data['dataAplicacao'])
 
         return data
+
+    def _validate_agrotoxico_xor_fotoAgrotoxico(self, agrotoxico, fotoAgrotoxico):
+        if (agrotoxico is not None) ^ (fotoAgrotoxico is not None):
+            return
+
+        raise serializers.ValidationError(
+            'É necessário enviar ou o agrotóxico ou a foto do agrotóxico.'
+        )
+
+    def _validate_agrotoxico(self, agrotoxico, plantio, dataAplicacao):
+        if not agrotoxico:
+            return
+
+        query = AplicacaoAgrotoxico.objects.filter(
+                plantio=plantio, agrotoxico=agrotoxico, dataAplicacao=dataAplicacao)
```

Comment:
Utilizando o models.Q do django
```suggestion
        query = AplicacaoAgrotoxico.objects.filter(
                Q(agrotoxio__isnull=False) & Q(agrotoxico=agrotoxico),
                plantio=plantio, dataAplicacao=dataAplicacao
         )
```

O bom dessa queria acima é que ela seria realizada toda a nível de banco antes de chegar no nível de aplicação
🐱‍👤😹🙀

Reply:
Não testei pra ver se isso funcionava

Reply:
Acho que não tem necessidade dessa otimização, visto que a verificação os dados agrotoxico, plantio e dataAplicaçao já estão na memoria. No final, daria no mesmo, ou até pior, se o agrotoxico for null. Acho que pode ficar como debito técnico e ser analisado depois

Reply:
Sem problemas por mim